DAOS-6515 placement: Find spare targets even when fault domains are used

byron-marohn opened the pull request:
Prior to this patch, placement would refuse to place a failed shard if
there aren't enough available top-level fault domains to guarantee
redundancy (for example, 4-way replication, and one rank fails). While
is a nice idea to try to encourage data safety, at the point where a
failure occurs data safety is already at risk. Refusing to rebuild that
data onto a still-living spare target does more harm than good.

This patch enables jump map placement to always find a spare target so
long as unused targets are available in the system, even if that means
that the redundancy available is less than the object class requires.

Signed-off-by: Byron Marohn <byron.marohn@intel.com>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ryon-jensen reviewed (approved):
works for me :)

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

liuxuezhao commented:
For the data verification failure, I think it has nothing to do with this PR.
It is a EC aggregation's bug, there is a related ticket https://jira.hpdd.intel.com/browse/DAOS-6755 I'll push the fix very recently this week.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

byron-marohn commented:
@ryon-jensen @wangdi1 @NiuYawei 

Re-review please. Only change since last review was skip() the DEGRADE4 test which has another issue that this patch just seems to trigger more often. Ticket created for re-enabling that test is DAOS-6991. 

I skipped Hardware small and Hardware large testing on this PR and only enabled some of the medium tests - as both of the previous two runs passed everything except DEGRADE4, which is run as part of Hardware Medium testing.
